core/src/main/java/org/elasticsearch/common/collect/CopyOnWriteHashMap.java

```diff
@@ -511,12 +516,12 @@ public int size() {
      * Same as {@link #copyAndPut(Object, Object)} but for an arbitrary number of entries.
      */
     public CopyOnWriteHashMap<K, V> copyAndPutAll(Map<? extends K, ? extends V> other) {
-        return copyAndPutAll(other.entrySet());
+        return copyAndPutAll(other.entrySet().stream());
     }
 
-    <K1 extends K, V1 extends V> CopyOnWriteHashMap<K, V> copyAndPutAll(Collection<Map.Entry<K1, V1>> entries) {
+    public <K1 extends K, V1 extends V> CopyOnWriteHashMap<K, V> copyAndPutAll(Stream<Entry<K1, V1>> entries) {
         CopyOnWriteHashMap<K, V> result = this;
-        for (Map.Entry<K1, V1> entry : entries) {
+        for (Entry<K1, V1> entry : ((Iterable<Entry<K1, V1>>)() -> (entries.iterator()))) { // look at how ugly it is...
```

Comment:
What do you think of this?

    public <K1 extends K, V1 extends V> CopyOnWriteHashMap<K, V> copyAndPutAll(Iterable<Entry<K1, V1>> entries) {
        CopyOnWriteHashMap<K, V> result = this;
        for (Entry<K1, V1> entry : entries) {
            result = result.copyAndPut(entry.getKey(), entry.getValue());
        }
        return result;
    }

    public <K1 extends K, V1 extends V> CopyOnWriteHashMap<K, V> copyAndPutAll(Stream<Entry<K1, V1>> entries) {
        return copyAndPutAll(entries::iterator);
    }

It's not necessary but the call site on line 519 can then change to not call `Collection#stream` but the rest of the call sites can remain as is.